runtime/interpreter/interpreter_expression.go

```diff
@@ -1180,8 +1180,11 @@ func (interpreter *Interpreter) VisitCastingExpression(expression *ast.CastingEx
 
 	switch expression.Operation {
 	case ast.OperationFailableCast, ast.OperationForceCast:
-		valueStaticType := value.StaticType(interpreter)
-		valueSemaType := interpreter.MustConvertStaticToSemaType(valueStaticType)
+		// if the value itself has a mapped entitlement type in its authorization
+		// (e.g. if it is a reference to `self` or `base`  in an attachment function with mapped access)
+		// substitution must also be performed on its entitlements
+		valueSemaType := interpreter.substituteMappedEntitlements(interpreter.MustSemaTypeOfValue(value))
+		valueStaticType := ConvertSemaToStaticType(interpreter, valueSemaType)
```

Comment:
Why the `substituteMappedEntitlements` is specifically needed only during casting?
Is this something that needs to/could be done generically in all places (for e.g: inside `IsSubTypeOfSemaType`)?

Reply:
There are specific semantics for when an entitlement mapping appears in a casting expression, as this is the only place in the language where mappings can traverse the "boundary" between static and runtime behavior:

```
access(mapping M) fun foo(): auth(mapping M) &Int {
    return &self.intField as auth(mapping M) &Int
}
```

for example, needs to "instantiate" the `auth(mapping M) &Int` that appears in the cast to the concrete entitlements that it actually should have at runtime based on the entitlements used to access the method. This is what `substituteMappedEntitlements` does. Applying this more generally is not possible since there are places where we don't want entitlement mappings to be replaced with concrete ones (e.g. generally during subtyping or determining interface conformance).

Reply:
Great explanation! Could you please extend the comment in the code with a variant of it?